### PR TITLE
Added infrastructure to enable easy enrichment of GenTraversables.

### PR DESCRIPTION
**Pull request by milessabin**
As discussed on [scala-internals](https://groups.google.com/d/topic/scala-internals/XRbObRJvOkU/discussion). Sample usage [here](https://gist.github.com/2500326). Motivating blog post [here](http://yz.mit.edu/wp/true-scala-complexity/).

Builds successfully and passes all tests.

### CODE REVIEW

**Comment by adriaanm** [comment]
A IsElemOf Repr?

**Comment by milessabin** [comment]
Maybe something more literal minded would be better. Elem[A, Repr] is an alias for Repr => GenTraversableLike[A, Repr], so how about ToGenTraverasbleLike[A, Repr]? It's a bit more verbose, but OTOH it's descriptive.

**Comment by adriaanm** [comment]
Repr HasElem A ?

**Comment by milessabin** [comment]
That works for me. I'll update the pull request ...

**Comment by milessabin** [comment]
Renamed Elem[A, Repr] to HasElem[Repr, A] as suggested by @adriaanm. Rebased to master, tested and squashed. Also got rid of the extraneous newline change.

File: src/library/scala/collection/generic/package.scala [hunk 1]
```diff
@@ -4,6 +4,12 @@ import generic.CanBuildFrom
 package object generic {
   type CanBuild[-Elem, +To] = CanBuildFrom[Nothing, Elem, To]
 
+  /** The type of conversions from a collection representation type
+   *  `Repr` to its corresponding GenTraversableLike.
+   *  @see FromRepr
```
Comment (retronym) [inline src/library/scala/collection/generic/package.scala]: `@see [[scala.collection.generic.FromRepr]]`

File: src/library/scala/collection/generic/FromRepr.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  implicit val stringFromRepr = new FromRepr[String] {
+    type A = Char
+    val hasElem = implicitly[HasElem[String, Char]]
+  }
+
+  implicit def genTraversableLikeFromRepr[C[_], A0]
+    (implicit hasElem0: HasElem[C[A0], A0]) = new FromRepr[C[A0]] {
+      type A = A0
+      val hasElem = hasElem0
+    }
```
Comment (retronym) [inline src/library/scala/collection/generic/FromRepr.scala]: Perhaps best to annotate the return types, e.g:

``` scala
implicit val stringFromRepr: FromRepr[String] { type A = Char} = new FromRepr[String] {
```

File: src/library/scala/collection/GenTraversableLike.scala [hunk 3]
```diff
@@ -411,3 +411,12 @@ trait GenTraversableLike[+A, +Repr] extends Any with GenTraversableOnce[A] with
   def stringPrefix: String
 
 }
+
+object GenTraversableLike {
+  /** Manufacture a conversion from collection representation type `Repr` to
+   *  its corresponding `GenTraversableLike` given an implicitly available
+   *  instance of `FromRepr[Repr]`.
+   *  @see FromRepr
```
Comment (retronym) [inline src/library/scala/collection/GenTraversableLike.scala]: as below...

Comment (milessabin) [inline src/library/scala/collection/GenTraversableLike.scala]: Also as below ... ;-)

File: src/library/scala/collection/generic/package.scala (hunk 1 above)
Comment (milessabin) [inline src/library/scala/collection/generic/package.scala]: Good catch ... thanks.

**Comment by adriaanm** [comment]
Thanks!
sorry, you'll need to submit a new pull-request for the buildbot to see it

File: src/library/scala/collection/generic/FromRepr.scala (hunk 2 above)
Comment (milessabin) [inline src/library/scala/collection/generic/FromRepr.scala]: As a general rule I find type annotations with refinements offensive non-DRY, but given that this is an implicit definition, I have to agree.

**Comment by milessabin** [comment]
Okey doke ... new pull request on its way. I'm closing this one.

**Pull request closed without merging**